string type
uint8 component
bool direction
